pallets/charity/src/tests.rs

```diff
@@ -186,6 +207,17 @@ fn allocating_works() {
 		// Check that the correct event is emitted
 		let new_pot_total = Balances::minimum_balance() + donation - alloc;
 		let expected_event = TestEvent::charity(RawEvent::FundsAllocated(2, 5, new_pot_total));
+
+		// testing if the the event come in the correct order
+		assert_eq!(
+			System::events()[3],
```

Comment:
This is interesting. Why is this coming as event #3?  If there are other legitimately-expected events before it, we should have asserts for them as well. I'm guessing those events are coming from the balances pallet. In that case we should test for them or at minimum make a comment about what they are and why ours its third.

Reply:
This is the list of the events that we get when the System::events() gets called.

- [EventRecord { phase: Phase::Initialization, event: TestEvent::balances(RawEvent::Transfer(1, 8241983431855337325, 10)), topics: [] }, 

- EventRecord { phase: Phase::Initialization, event: TestEvent::charity(RawEvent::DonationReceived(1, 10, 11)), topics: [] }, 
- EventRecord { phase: Phase::Initialization, event: TestEvent::balances(RawEvent::Transfer(8241983431855337325, 2, 5)), topics: [] }, 
- EventRecord { phase: Phase::Initialization, event: TestEvent::charity(RawEvent::FundsAllocated(2, 5, 6)), topics: [] }]

Reply:
Based on the comments above, what should I do in this case?

Reply:
Great, that makes sense. Only some of those events are coming from the charity pallet. Here's what the four events are:

0. From balances pallet. It is from the tokens we transferred when we were first making the donation.
1. From the charity pallet confirming the donation was received.
2. From the balances pallet. It is another transfer when the funds are allocated.
3. From the charity pallet confirming the allocation succeeded.

All of these events are relevant to our pallet, so we should test for all of them. You could do that with four separate `assert_eq!`s, or with one big one like this:

```rust
assert_eq!(
  System::events(),
  vec![
    TestEvent::balances(...),
    TestEvent::charity(...),
    TestEvent::balances(...),
    TestEvent::charity(...),
  ]
);
```

Reply:
For different tests, we do receive different list of events: 
imbalances_work test : 
[EventRecord { phase: Phase::Initialization, event: TestEvent::charity(RawEvent::ImbalanceAbsorbed(5, 6)), topics: [] }]

donations_work test: 
[EventRecord { phase: Phase::Initialization, event: TestEvent::balances(RawEvent::Transfer(1, 8241983431855337325, 10)), topics: [] }, 
EventRecord { phase: Phase::Initialization, event: TestEvent::charity(RawEvent::DonationReceived(1, 10, 11)), topics: [] }]

allocationg_works test: 
[EventRecord { phase: Phase::Initialization, event: TestEvent::balances(RawEvent::Transfer(1, 8241983431855337325, 10)), topics: [] }, 
EventRecord { phase: Phase::Initialization, event: TestEvent::charity(RawEvent::DonationReceived(1, 10, 11)), topics: [] }, 
EventRecord { phase: Phase::Initialization, event: TestEvent::balances(RawEvent::Transfer(8241983431855337325, 2, 5)), topics: [] }, 
EventRecord { phase: Phase::Initialization, event: TestEvent::charity(RawEvent::FundsAllocated(2, 5, 6)), topics: [] }]

So we can not test all the four events inside the donations_work test because the System::events(); gives us only two events.

Reply:
Yes, each test simulates different transactions, so of course you will receive different events. `imbalances_work` emits one event and you should test for it. `donations_work` emits two events and you should test for both of them. `allocationg_works` emits four events and you should test for all four of them as I showed above.

Reply:
One problem that I am facing is this one: 
![image](https://user-images.githubusercontent.com/13895430/91614488-efb6a980-e981-11ea-8fb5-13792aa302bc.png)

Reply:
I've pushed a commit that shows how to make this test work. But now that I've thought about it more, it might not really make sense to test for other pallets' events. Maybe we should get all the events, then filter to just the ones related to our pallet, then make sure they come in the correct order. What do you think @herou

Reply:
Well, let's assume we are inside the allocating_works function test. Since we do get other pallet events we have to test them as well even though those events are not part of our pallet. If we do as you are mentioning above, we will make things more complicated and every test would be "unique" and different from the other tests and would not be easy to maintain. It is up to you how would you like to proceed. If you are ok with the "old" way, let me know so I can work on the other tests. @JoshOrndorff

Reply:
@JoshOrndorff Any response?  I am looking for another ticket (which has the label mentor available) to work on.

Reply:
Hi @herou 

I didn't realize you were waiting on me. I looked through the code in the main Substrate repo, and found a good pattern for testing events. I've updated the Charity and compounding interest pallets to use this pattern. I think it is the best of both worlds. We are testing the order of events, and only worrying about our own pallets events.

Please take a look at the commits I added, and ask if you have any questions.

Reply:
@JoshOrndorff Ok, It is clear for me. Do you want to proceed like this for all the tests? I would appreciate if you can reply to me more often. Thank you.

Reply:
Yes, please proceed this way for the other tests :+1: